lib/DB/Schema/ResultSet/Course.pm

```diff
@@ -262,36 +335,202 @@ This gets the Course Settings for a course
+
+	my $global_setting = $self->getGlobalSetting(info => getSettingInfo($args{info}));
+
+	my $course_setting = $course->course_settings->find({
+		setting_id => $global_setting->{setting_id}
+	});
 
-	my $current_settings = { $course->course_settings->get_inflated_columns };
-	my $updated_settings = mergeCourseSettings($current_settings, $args{settings});
+	# Check that the setting is valid.
 
-	my $cs = $course->course_settings->update($updated_settings);
-	return mergeCourseSettings(getDefaultCourseValues(), { $cs->get_inflated_columns });
+	my $params = {
+		course_id  => $course->course_id,
+		setting_id => $global_setting->{setting_id},
+		value      => { value => $args{params}{value} }
+	};
+
+	# remove the following fields before checking for valid settings:
+	delete $global_setting->{$_} for (qw/setting_id course_id/);
+
+	isValidSetting($global_setting, $params->{value}{value});
+
+	# The course_id must be deleted to ensure it is written to the database correctly.
+	delete $params->{course_id} if defined($params->{course_id});
+
+	my $updated_course_setting =
+		defined($course_setting) ? $course_setting->update($params) : $course->add_to_course_settings($params);
+
+	return $updated_course_setting if $args{as_result_set};
+	my $setting_to_return =
+		($args{merged})
+		? { $updated_course_setting->get_inflated_columns,
+			$updated_course_setting->global_setting->get_inflated_columns }
+		: { $updated_course_setting->get_inflated_columns };
+
+	# value and default_value are decoded from JSON as a hash.  Return to a value.
+	for my $key (qw/default_value value/) {
+		$setting_to_return->{$key} = $setting_to_return->{$key}->{value} if defined($setting_to_return->{$key});
+	}
+	return $setting_to_return;
+}
+
+=pod
+
+=head2 deleteCourseSetting
+
+Delete a single course setting.
+
+=head3 input
+
+=over
+
+=item * C<info> which is a hash containing information about the course (either a
+C<course_id> or C<course_name>) and a setting (either a C<setting_id> or C<setting_name>).
+
+=item * C<$as_result_set>, a boolean if the return is to be a result_set
+
+=back
+
+=head3 output
+
+A single course setting as either a hashref or a C<DBIx::Class::ResultSet::CourseSetting> object.
+
+=cut
+
+sub deleteCourseSetting ($self, %args) {
+	my $setting         = $self->getCourseSetting(info => $args{info}, as_result_set => 1);
+	my $deleted_setting = $setting->delete;
+
+	return $deleted_setting if $args{as_result_set};
+
+	my $setting_to_return =
+		($args{merged})
+		? { $deleted_setting->get_inflated_columns, $deleted_setting->global_setting->get_inflated_columns }
+		: { $deleted_setting->get_inflated_columns };
+
+	# value and default_value are decoded from JSON as a hash.  Return to a value.
+	for my $key (qw/default_value value/) {
+		$setting_to_return->{$key} = $setting_to_return->{$key}->{value} if defined($setting_to_return->{$key});
+	}
+	return $setting_to_return;
```

Comment:
I thought we discussed *not* returning objects from deletion calls...

Reply:
I wrote this before the deleteXXX PR.  I can update.

Reply:
done.